src/App.jsx

```diff
@@ -1,65 +1,132 @@
-    ]
-    const initialBlogs = initialTitles.map(title => titleToBlog(title))
-    const [blogTitles, setBlogTitles] = useState([])
-    const [isLoading, setIsLoading] = useState(true)
+  const initialBlogs = initialTitles.map(title => titleToBlog(title));
+  const [state, runAction] = useReducer(stateManagementFunction, initialState);
 
-    useEffect(() => {
-        new Promise(
-            (resolve, reject) => {
-                setTimeout(() => {
-                    resolve(getItem(LOCALSTORAGE_KEY) || initialBlogs)
-                }, 2000)
-            }
-        ).then((blogTitlesFromLocalStorage) => {
-            setBlogTitles(blogTitlesFromLocalStorage)
-            setIsLoading(false)
-        })
-    }, [])
+  useEffect(() => {
+    setTimeout(() => {
+      const blogTitlesFromLocalStorage =
+        getItem(LOCALSTORAGE_KEY) || initialBlogs;
+      runAction({
+        type: 'FINISH_LOADING_BLOG_TITLES',
+        payload: { blogTitles: blogTitlesFromLocalStorage },
+      });
+    }, 2000);
+  }, []);
 
-    useEffect(() => {
-        !isLoading && setItem(LOCALSTORAGE_KEY, blogTitles)
-    }, [blogTitles])
+  useEffect(() => {
+    !state.isLoading && setItem(LOCALSTORAGE_KEY, state.blogTitles);
+  }, [state.blogTitles]);
 
-    const [searchText, setSearchText] = useState('');
-    const [focus, setFocus] = useState('search');
+  // focus needs to be reset after the latest focus is applied
+  useEffect(() => {
+    !!state.focus && runAction({ type: 'RESET_FOCUS' });
+  }, [state.focus]);
 
-    // focus needs to be reset after the latest focus is applied
-    useEffect(() => {
-        !!focus && setFocus()
-    }, [focus])
+  const onAddBlogTitle = newTitle => {
+    runAction({
+      type: 'START_ADDING_BLOG_TITLE',
+    });
+    setTimeout(() => {
+      runAction({
+        type: 'FINISH_ADDING_BLOG_TITLE',
+        payload: { newTitle: newTitle },
+      });
+    }, 2000);
+  };
 
-    const onAddBlogTitle = (newTitle) => {
-        setBlogTitles([...blogTitles, titleToBlog(newTitle)])
-        setFocus('addBlog')
-    }
-    return (
-        <div>
-            <h1>Cockatoo Lesson Blog Title</h1>
-            <AddBlogTitle focus={focus === 'addBlog'} onAddTitle={onAddBlogTitle}/>
-            <Search onSearch={setSearchText} focus={focus === 'search'}/>
-            {isLoading ? <p>Loading...</p> :
-                (<BlogTitles titles={blogTitles} searchText={searchText}/>)}
+  const onSearch = searchText =>
+    runAction({ type: 'SET_SEARCH_TEXT', payload: { searchText } });
 
-        </div>
-    );
+  return (
+    <div>
+      <h1>Cockatoo Lesson Blog Title</h1>
+      <AddBlogTitle
+        focus={state.focus === 'addBlog'}
+        onAddTitle={onAddBlogTitle}
+      />
+      <Search onSearch={onSearch} focus={state.focus === 'search'} />
+      {state.isLoading ? (
+        <p>Loading...</p>
+      ) : (
+        <BlogTitles
+          titles={filterBlogTitles(state.blogTitles, state.searchText)}
```

Comment:
I moved the blog filtering logic outside of the BlogTitles component since the logic better belongs in App